### ISSUE TITLE
Potential error when trying to fail transactions early

### ISSUE THREAD

**Issue by LiangGeorge** [XSEDE/amieclient#30]
To my knowledge, packets enforce a type checking of packet it expects as a reply. When I found a packet in the packet sequence (from the test API) that I wanted to fail earlier (because of missing information) with the reply_with_failure() method, I was given this error

`File "/lustre/software/anaconda3/aarch64/envs/AMIE/lib/python3.7/site-packages/amieclient-0.6.0-py3.7.egg/amieclient/client.py", line 223, in send_packet
amieclient.client.AMIERequestError: Packet type not in expected replies`

Are packets supposed to be failed at the last stage where it expects ITC? Or can they be failed earlier?

Thanks,
George

**Comment by sabo** [comment]
Hi George, can you give me some info on what specific type of packet you were replying to? reply_with_failure() should give you an ITC, and I believe you're right that an ITC packet with a failure should be accepted by the API at any stage.

Also adding @ericblau: given [the location](https://github.com/XSEDE/amieclient/blob/743c3c7dbecd2910c836b24391d67f3b0399c803/amieclient/client.py#L221-L223) generating the error, this message is coming from the server side of the API

https://github.com/XSEDE/amieclient/blob/743c3c7dbecd2910c836b24391d67f3b0399c803/amieclient/client.py#L221-L223

**Comment by LiangGeorge** [comment]
Hello Sabo, 
I'm trying to reply to a request_project_create packet. 
The line of code that I wrote looks like this once I've determined I want to fail a packet. ` amie_client.send_packet(packet.reply_with_failure())`

**Comment by ericblau** [comment]
I agree w/ Ryan that you are supposed to be able to fail a transaction with the right ITC packet at any point.  I will look into the server side implementation to see if I can tell what is going on.

**Comment by LiangGeorge** [comment]
As a follow-up, I tried to work around the issue detailed above by failing packets when we expect a ITC packet as a reply and ran into another issue. For a single trans_rec_id it went from "request_project_create" to "data_project_create" (which has an expected type of "inform_transaction_complete"). 

I have a function that returns the itc, and another one that sends it: 
```
fail = user_prompt.yn_prompt("Is the current transaction one that we want to fail? [Y/n]")
    if fail == 'Y':
        print("Failing packet")
        itc = packet.reply_with_failure()
        #I assume the below is optional because of the method above, but I set the fields anyway
        itc.StatusCode = 'Failed'
        itc.DetailCode = '2'
        itc.Message = "Failing Transaction"
        return itc
```

Code that sends the itc returned above:
```
itc = packet_helper.process_dpc(packet)
    # send the ITC
    amie_client.send_packet(itc)
```

I get the following error:
```
Failing packet
`Traceback` (most recent call last):
  File "process_packets.py", line 122, in <module>
    amie_client.send_packet(itc)
  File "/lustre/software/anaconda3/aarch64/envs/AMIE/lib/python3.7/site-packages/amieclient-0.6.0-py3.7.egg/amieclient/client.py", line 223, in send_packet
amieclient.client.AMIERequestError: PG::RaiseException: ERROR:  StatusCode(Failed) not valid
CONTEXT:  PL/pgSQL function incoming_inform_transaction_complete(integer) line 26 at RAISE
SQL statement "select amie.incoming_inform_transaction_complete(222154314);"
PL/pgSQL function packet_dispatcher() line 43 at `EXECUTE`

```

To reiterate I'm using the testing API so I'm not sure if failed transactions are allowed with it.

**Comment by ericblau** [comment]
Looking at the amie_api server-side code, and the amieclient code, it appears that amieclient reply_with_failure() method is sending an ITC with status code "Failed" whereas the amie_api expects "Failure" as the status code.  I have confirmed with Michael Shapiro that the underlying AMIE implementation expects "Failure" in this context, so this is a bug in the amieclient code.

**Comment by ericblau** [comment]
StatusCode for an ITC can only be "Success" or "Failure", so I'm not surprised by strange errors if it is anything else.

**Comment by sabo** [comment]
Just fixed this, will fix one more small issue (#29 ) and push out a new release by the end of the day
